File: Numbers/215-Kth-Largest-Element-in-an-Array.py
from heapq import (
    heappop, heappush
    )
import random
class Solution(object):
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        https://leetcode.com/problems/kth-largest-element-in-an-array/
        """
        #quick selection algorithm O(n)
        #randomizing
        for i in xrange(len(nums)):
            ran = random.randint(0, len(nums)-1)
            nums[i], nums[ran] = nums[ran], nums[i]
            
        start, stop = 0, len(nums)-1
        target, pi = len(nums)-k, len(nums)-1
        pi = self.findPivot(start, stop, nums)
        while target != pi:
            if pi > target:
                stop = pi - 1
                pi = self.findPivot(start, stop, nums)
            else:
                start = pi + 1
                pi = self.findPivot(start, stop, nums)
        return nums[target]
        
    def findPivot(self, start, stop, nums):
        """find the pivot,
        #@ return type int
        """
        if start == stop:
            return start
        pi = start
        pivot = stop
        for i in xrange(start, stop):
            #print nums[i]
            if nums[i] <= nums[pivot]:
                nums[i], nums[pi] = nums[pi], nums[i]
                pi += 1
        nums[pivot], nums[pi] = nums[pi], nums[pivot]
        #print "pi is {}".format(pi)
        return pi
        
        
        
        
        
        
        #solution2 O(nlogk)
        #h = []
        #for num in nums:
        #    heappush(h, num)
        #    if len(h) > k:
        #        heappop(h)
        #return heappop(h)
        
        #solution1 O(nlogn)
        #nums.sort()
        #return nums[len(nums)-k]

File: Numbers/215-Kth-Largest-Element-in-an-Array.java
public class Solution {
    public int findKthLargest(int[] nums, int k) {
        int start = 0;
        int last = nums.length - 1;
        int target = nums.length - k;
        int pivot = findPivot(0, last, nums); 
        while (target != pivot){
            if (target < pivot){
                pivot = findPivot(0, pivot-1, nums);
            }else{
                pivot = findPivot(pivot+1, last, nums);
            }
        }
        return nums[target];
        
    }
    
    private int findPivot(int start, int end, int[] nums){
        int pPivot = end;
        int pi = start;
        for (int i = start; i < end; i++){
            if (nums[i] <= nums[pPivot]){
                int tmp = nums[i];
                nums[i] = nums[pi];
                nums[pi] = tmp;
                pi++;
            }
        }
        int tmp = nums[pi];
        nums[pi] = nums[pPivot];
        nums[pPivot] = tmp;
        return pi;
    }
}
